include/systems/implicit_system.h

```diff
@@ -405,82 +308,13 @@ class ImplicitSystem : public ExplicitSystem
   bool zero_out_matrix_and_rhs;
 
 protected:
-
-  /**
-   * Initializes the member data fields associated with
-   * the system, so that, e.g., \p assemble() may be used.
-   */
-  virtual void init_data () override;
-
-  /**
-   * Initializes the matrices associated with this system.
-   */
-  virtual void init_matrices ();
-
-private:
-
   /**
-   * Add the system matrix to the \p _matrices data structure.
-   * Useful in initialization.
+   * Adds the system matrix
    */
-  void add_system_matrix ();
-
-  /**
-   * Some systems need an arbitrary number of matrices.
-   */
-  std::map<std::string, SparseMatrix<Number> *> _matrices;
-
-  /**
-   * Holds the types of the matrices
-   */
-  std::map<std::string, ParallelType> _matrix_types;
-
-  /**
-   * \p true when additional matrices may still be added, \p false otherwise.
-   */
-  bool _can_add_matrices;
+  virtual void add_matrices() override;
 };
 
 
-
-// ------------------------------------------------------------
-// ImplicitSystem inline methods
-inline
-bool ImplicitSystem::have_matrix (const std::string & mat_name) const
-{
-  return (_matrices.count(mat_name));
-}
-
-
-inline
-unsigned int ImplicitSystem::n_matrices () const
-{
-  return cast_int<unsigned int>(_matrices.size());
-}
-
-template <template <typename> class MatrixType>
-inline
-SparseMatrix<Number> &
-ImplicitSystem::add_matrix (const std::string & mat_name,
-                            const ParallelType type)
-{
-  // only add matrices before initializing...
-  libmesh_error_msg_if(!_can_add_matrices,
-                       "ERROR: Too late.  Cannot add matrices to the system after initialization"
-                       "\n any more.  You should have done this earlier.");
-
-  // Return the matrix if it is already there.
-  if (this->have_matrix(mat_name))
-    return *(_matrices[mat_name]);
-
-  // Otherwise build the matrix and return it.
-  SparseMatrix<Number> * buf = libmesh_make_unique<MatrixType<Number>>(this->comm()).release();
-  _matrices.emplace(mat_name, buf);
-  _matrix_types.emplace(mat_name, type);
-
-  return *buf;
-}
```

Comment:
I love Github's diff viewer but I'd love it a lot more if it displayed code movement with more information than just "something was deleted here and also something (else?) was added there".  Did you make any changes in these moves other than the necessary ImplicitSystem->System syntax edits?

Reply:
All of the standard getters (get_matrix, n_matrices, have_matrix, checking if we can add matrices or not) were just moved copy-pasta. This also includes all of the doc strings in the headers.

The init_data() and init_matrices() methods for ImplicitSystem were absorbed into the init for System, with the (new) remaining piece put into add_matrices(). Both ImplicitSystem and EigenSystem override add_matrices to define their respective matrices.